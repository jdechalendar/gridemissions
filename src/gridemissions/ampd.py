"""
Tools to interact with the AMPD ftp at newftp.epa.gov
"""
import os
from os.path import join
import pandas as pd
from ftplib import FTP
import logging


class AMPD_download(object):
    def __init__(self, pathOut):
        self.ftp = FTP("newftp.epa.gov")
        os.makedirs(pathOut, exist_ok=True)
        self.pathOut = pathOut
        self.logger = logging.getLogger("scraper")
        self.localStatus = {}
        self.remoteStatus = {}

    def login(self):
        self.ftp.login()

    def close(self):
        self.ftp.close()

    def getStatus(self, year=None):
        self.ftp.cwd("/dmdnload/emissions/hourly/monthly")
        self.years = self.ftp.nlst()

        if year is not None:
            years = [str(year)]
        else:
            years = self.years

        for y in years:
            self.logger.info("Retrieving ftp status for %s" % y)
            self.localStatus[y] = self.getLocalStatus(y)
            self.remoteStatus[y] = self.getRemoteStatus(y)
            self.logger.info("%s has %d files" % (y, len(self.remoteStatus[y])))

    def getLocalStatus(self, year):
        """
        Get status of AMPD dataset on local filesystem:
        What files are available.
        TODO: store and reload what the modification date was the last
        time we downloaded stuff.
        """
        folderYear = join(self.pathOut, year)
        os.makedirs(folderYear, exist_ok=True)
        return pd.DataFrame([[f] for f in os.listdir(folderYear)], columns=["fileNm"])

    def getRemoteStatus(self, year):
        """
        Get status of AMPD dataset on EPA ftp:
        What files are available, what modification date.
        """
        self.ftp.cwd("/dmdnload/emissions/hourly/monthly/%s" % year)
        fileNms = self.ftp.nlst()

        fileNms = [(self.ftp.sendcmd("MDTM %s" % f), f) for f in fileNms]
        df = pd.DataFrame(fileNms, columns=["MDTM", "fileNm"])
        df.MDTM = df.MDTM.apply(
            lambda x: pd.to_datetime(x.split()[-1], format="%Y%m%d%H%M%S.%f")
        )
        return df

    def download(self, year):
        """
        Download a year's worth of data, skipping files that are already there.
        """
        year = str(year)
        self.logger.info("Downloading data for %s" % year)
        folderYear = join(self.pathOut, year)

        self.ftp.cwd("/dmdnload/emissions/hourly/monthly/%s" % year)

        cnt_downloaded = 0
        for irow, row in self.remoteStatus[year].iterrows():
            # Check to see if we already have this
            local = self.localStatus[year][self.localStatus[year].fileNm == row.fileNm]

            if len(local) == 0:
                download = True
            elif len(local) == 1:
                download = False
            else:
                logging.debug(local)
                raise ValueError("There are two fileNms that match %s" % row)

            if download:
                try:
                    self.logger.debug("Downloading %s ..." % row.fileNm)
                    with open(join(folderYear, row.fileNm), "wb") as f:
                        self.ftp.retrbinary("RETR %s" % row.fileNm, f.write)
                    cnt_downloaded += 1
                except TimeoutError:  # TODO add known errors here
                    return "timedout"
            else:
                self.logger.debug("Skipping %s ..." % row.fileNm)
        self.logger.info(
            "Downloaded  %d/%d files for %s"
            % (cnt_downloaded, len(self.remoteStatus[year]), year)
        )
        return "done"


class AMPDRaw(object):
    """
    Simple class to handle AMPD data.
    """

    def __init__(self, fileNms):
        self.logger = logging.getLogger("load")
        if isinstance(fileNms, str):
            fileNms = [fileNms]
        self.fileNms = fileNms

    def parse(self, cols=None):
        self.df = pd.concat([self._parse_file(f, cols) for f in self.fileNms])

    def _parse_file(self, fileNm, cols=None):
        """
        Parse one file downloaded from the EPA AMPD ftp server
            ftp://newftp.epa.gov/DMDnLoad/emissions/hourly/monthly/

        1. Create a datetime column.
        2. Change units to metric
        3. Select only columns we need
        """
        col_name_map = {
            "CO2_MASS": "CO2_MASS (tons)",
            "CO2_RATE": "CO2_RATE (tons/mmBtu)",
            "GLOAD": "GLOAD (MW)",
            "HEAT_INPUT": "HEAT_INPUT (mmBtu)",
            "NOX_MASS": "NOX_MASS (lbs)",
            "SLOAD": "SLOAD (1000lb/hr)",
            "SLOAD (1000 lbs)": "SLOAD (1000lb/hr)",
            "SO2_MASS": "SO2_MASS (lbs)",
            "SO2_RATE": "SO2_RATE (lbs/mmBtu)",
            "NOX_RATE": "NOX_RATE (lbs/mmBtu)",
        }
        us_to_si_tons = 0.9071847
        lbs_to_si_tons = 0.000453592
        mmbtu_to_mwh = 0.29329722222222

        # Read data
        self.logger.info("Loading %s" % fileNm)
        df_tmp = pd.read_csv(fileNm, compression="zip", low_memory=False)

        # Rename columns
        df_tmp.rename(col_name_map, axis="columns", inplace=True)

        # 1. Parse timestamp (local time)
        df_tmp.loc[:, "OP_DATE_TIME"] = pd.to_datetime(
            df_tmp["OP_DATE"] + "-" + df_tmp["OP_HOUR"].astype(str),
            format="%m-%d-%Y-%H",
        )
        #        df_tmp.loc[:, 'YEAR'] = df_tmp.loc[:, 'OP_DATE_TIME'].dt.year.astype(
        #            int)
        #        df_tmp.loc[:, 'MONTH'] = df_tmp.loc[:, 'OP_DATE_TIME'].dt.month.astype(
        #            int)

        # 2. Change units to metric
        df_tmp.loc[:, "CO2"] = df_tmp.loc[:, "CO2_MASS (tons)"] * us_to_si_tons
        df_tmp.loc[:, "SO2"] = df_tmp.loc[:, "SO2_MASS (lbs)"] * lbs_to_si_tons
        df_tmp.loc[:, "NOX"] = df_tmp.loc[:, "NOX_MASS (lbs)"] * lbs_to_si_tons
        df_tmp.loc[:, "HEAT_INPUT"] = df_tmp.loc[:, "HEAT_INPUT (mmBtu)"] * mmbtu_to_mwh
        df_tmp.loc[:, "GLOAD"] = df_tmp.loc[:, "GLOAD (MW)"]

        # 3. Keep as little info as possible to save space
        keep_cols = [
            "ORISPL_CODE",
            "UNITID",
            "OP_DATE_TIME",
            "OP_TIME",
            "GLOAD",
            "SO2",
            "NOX",
            "CO2",
            "HEAT_INPUT",
        ]

        if cols is None:
            return df_tmp.loc[:, keep_cols]
        elif cols == "all":
            return df_tmp
        else:
            return df_tmp.loc[:, cols]


def extract_state(x):
    """
    Combine all the files for a given state and year.
    """
    path_name, path_name_out, year, state = x
    fileNms = [
        join(path_name, f) for f in os.listdir(path_name) if f.startswith(year + state)
    ]
    logger = logging.getLogger("load")
    logger.info("Reading %d files for %s (%s)" % (len(fileNms), state, year))

    ampd = AMPDRaw(fileNms)
    ampd.parse()
    ampd.df.to_csv(join(path_name_out, "%s%s.csv" % (state, year)), index=False)


if __name__ == "__main__":
    from gridemissions import config

    # Download data for 2019
    DATA_PATH = config["DATA_PATH"]
    count_timed_out = 0
    status = "timedout"
    y = 2019
    while status == "timedout":
        ampd = AMPD_download(os.path.join(DATA_PATH, "raw", "AMPD"))
        ampd.login()
        ampd.getStatus(y)
        status = ampd.download(y)
        if status == "timed_out":
            count_timed_out += 1
        if count_timed_out > 5:
            raise ValueError("Too many timeouts")
        ampd.close()


#     # Extract data for California
#     state = "tx"
#     for year in ["2015", "2016", "2017", "2018", "2019"]:
#         path_in = join(config['DATA_PATH'], "raw", "AMPD", year)
#         path_out = join(config['DATA_PATH'], "analysis", "AMPD")
#         os.makedirs(path_out, exist_ok=True)
#         extract_state((path_in, path_out, year, state))
