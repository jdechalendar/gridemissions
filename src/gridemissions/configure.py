import json
import logging.config
import os
from pathlib import Path
import sys
from typing import Union

LOG_FORMAT = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"


def get_environ_variable(
    expected_variable_name: str, is_path: bool = False
) -> Union[Path, str, None]:
    """
    Get an environment variable.

    Optionally retrieves the variable as a path.
    Parameters
    ----------
    expected_variable_name: The expected environment variable name.
    is_path: Whether to try and interpret the variable as a path.

    Returns
    -------
    value: Union[Path, str, None]
        The value of the environment variable.
    """
    value = os.environ.get(expected_variable_name)

    if value:
        if is_path:
            return Path(value)

    return value


CONFIG_DIR_PATH = get_environ_variable(
    "GRIDEMISSIONS_CONFIG_DIR_PATH", is_path=True
) or Path.home().joinpath((".config/gridemissions"))
CONFIG_DIR_PATH.mkdir(exist_ok=True, parents=True)

LOG_CONFIG_FILE_PATH = get_environ_variable(
    "GRIDEMISSIONS_LOG_CONFIG_FILE_PATH", is_path=True
) or CONFIG_DIR_PATH.joinpath("logging.conf")
CONFIG_FILE_PATH = get_environ_variable(
    "GRIDEMISSIONS_CONFIG_FILE_PATH", is_path=True
) or CONFIG_DIR_PATH.joinpath("config.json")

try:
    with open(CONFIG_FILE_PATH, "r") as f:
        config = json.load(f)
except FileNotFoundError:
    print(f"Generating config file in {CONFIG_DIR_PATH} with default values")
    data_dir_path = get_environ_variable(
        "GRIDEMISSIONS_DATA_DIR_PATH", is_path=True
    ) or Path.home().joinpath("data/gridemissions")
    tmp_dir_path = get_environ_variable(
        "GRIDEMISSIONS_TMP_DIR_PATH", is_path=True
    ) or Path.home().joinpath("tmp/gridemissions")
    data_dir_path.mkdir(exist_ok=True, parents=True)
    tmp_dir_path.mkdir(exist_ok=True, parents=True)
    config = {
        "DATA_PATH": str(data_dir_path),
        "TMP_PATH": str(tmp_dir_path),
    }

    # Store for later use
    with open(CONFIG_FILE_PATH, "w") as f:
        json.dump(config, f, indent=4)

    # Sanity check: reload
    with open(CONFIG_FILE_PATH, "r") as f:
        config = json.load(f)

for k in config:
    if k.endswith("_PATH"):  # by convention, this is a directory
        config[k] = Path(config[k])

if "ENV" not in config:
    config["ENV"] = "PROD"

config["API_URL"] = "http://localhost:5000"

if config["ENV"] == "PROD":
    config["API_URL"] = "https://api.gridemissions.com"
config["S3_URL"] = "https://gridemissions.s3.us-east-2.amazonaws.com/"


def configure_logging(level="WARNING"):
    """
    Should be called by the user to configure logging

    Console logging is handled by the root logger's handler.
    To set the log level of the console to `console_log_level`, use
    ```
    for h in logging.getLogger().handlers:
        if isinstance(h, logging.StreamHandler):
            h.setLevel(console_log_level)
    ```
    """
    logging.basicConfig(format=LOG_FORMAT, level=level)
    logger = logging.getLogger()
    logger.setLevel(level)


if __name__ == "__main__":
    try:
        ret_val = config.get(sys.argv[1]) or ""
    except (KeyboardInterrupt, SystemExit):
        raise
    except:
        ret_val = ""

    sys.stdout.write("%s\n" % ret_val)
