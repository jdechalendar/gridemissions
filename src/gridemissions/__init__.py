import importlib.metadata

from .configure import config, configure_logging
from .load import GraphData, read_csv
from .emissions import EmissionsCalc

try:
    from .clean import BasicCleaner, RollingCleaner, CvxCleaner

    has_optional_dependencies = True

except ModuleNotFoundError:  # Cleaning depends on pptional dependencies
    has_optional_dependencies = False

__all__ = ["config", "configure_logging", "EmissionsCalc", "GraphData", "read_csv"]
__version__ = importlib.metadata.version("gridemissions")

if has_optional_dependencies:
    __all__ += ["BasicCleaner", "CvxCleaner", "RollingCleaner"]
