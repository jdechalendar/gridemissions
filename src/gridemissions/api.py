import io
import logging
from typing import List, Union
import requests
import pandas as pd

import gridemissions
from gridemissions import eia_api

DATETIME_FMT = "%Y%m%dT%H%MZ"
DATASET_TO_VARIABLE = {"raw": "E", "co2": "CO2", "elec": "E", "co2i": "CO2i"}

logger = logging.getLogger(__name__)


def retrieve(dataset, start=None, end=None, return_type="dataframe", **kwargs):
    """
    Retrieve data from the gridemissions API. Additional kwargs will be passed to
    GraphData.get_cols on the server side of the API.

    Parameters
    ----------
    dataset : str in DATASET_TO_VARIABLE.keys(). Currently, ["co2", "raw", "elec", "co2i"]
    start : str | datetime-like | None
        must be parseable by pd.to_datetime. Assumes UTC
    end : str | datetime-like | None
        must be parseable by pd.to_datetime. Assumes UTC
    return_type: str {"dataframe", "text"}

    Other Parameters
    ----------------
    region: str | List[str] | None
    field: str | List[str] | None
    region2: str | List[str] | None


    Notes
    -----
    For documentation of the "Other Parameters", refer to `GraphData.get_cols`.

    If `dataset == "co2i"`, the function behavior is slightly different. The API does
    not provide raw carbon intensity data, only consumption and production data, so in this
    case we make two calls to the API and recompute the carbon intensities. Current
    implementation only provides access to either consumption or production based carbon
    intensities. Column names are regions for which data are requested.

    If start or end are None, retrieves the last 24 hours of available data

    Examples
    --------

    To download co2 consumption data for CISO for the year of 2019:

    >>> from gridemissions import api
    >>> api.retrieve(dataset="CO2", region="CISO", start="20190101",
    ...              end="20200101", field="D")

    To download electricity generation data for ERCOT and BPAT for a year:

    >>> api.retrieve(dataset="E", region=["ERCOT", "BPAT"],
    ...              start="20190101", end="20200101", field="NG")
    """
    _check_arg_in_list(dataset, DATASET_TO_VARIABLE.keys())
    _check_arg_in_list(return_type, ["dataframe", "text"])

    if dataset == "co2i":
        if ("field" not in kwargs) or (kwargs["field"] not in ["D", "NG"]):
            raise ValueError("Dataset is co2i, field should be D or NG")

        co2 = retrieve(
            dataset="co2", start=start, end=end, return_type="dataframe", **kwargs
        )
        elec = retrieve(
            dataset="elec", start=start, end=end, return_type="dataframe", **kwargs
        )
        key_E = eia_api.get_key("E")
        key_CO2 = eia_api.get_key("CO2")
        key_CO2i = eia_api.get_key("CO2i")
        field = kwargs["field"]
        co2.columns = co2.columns.map(
            lambda x: eia_api.column_name_to_region(x, key_CO2[field])
        )
        elec.columns = elec.columns.map(
            lambda x: eia_api.column_name_to_region(x, key_E[field])
        )
        co2i = co2 / elec
        co2i.columns = co2i.columns.map(lambda x: key_CO2i[field] % x)

        if return_type == "dataframe":
            return co2i
        else:
            return co2i.to_csv()

    params = {"dataset": dataset}
    if (start is None) or (end is None):
        params["past24"] = "yes"
    else:
        params["start"] = pd.to_datetime(start).strftime(DATETIME_FMT)
        params["end"] = pd.to_datetime(end).strftime(DATETIME_FMT)
    for name in ["region", "field", "region2"]:
        if (name in kwargs) and (kwargs[name] is not None):
            params[name] = kwargs[name]

    url = gridemissions.config["API_URL"] + "/data"

    try:
        response = requests.get(url, params=params)
    except requests.exceptions.ConnectionError:
        logger.error(f"ConnectionError when connecting to {url}")
        if return_type == "dataframe":
            return pd.DataFrame()
        return ""

    if response.status_code == requests.codes.ok:
        if return_type == "dataframe":
            return pd.read_csv(
                io.StringIO(response.text), index_col=0, parse_dates=True
            )
        return response.text
    else:
        logger.error(f"{response.status_code} error: DEBUG for more info")
        logger.debug(f"{response.text}")
        if return_type == "dataframe":
            return pd.DataFrame()
        return ""


def _check_arg_in_list(
    arg: Union[str, List[str]], allowed_list: List[str], split: Union[str, None] = None
):
    """
    Helper function to sanitize inputs to the retrieve function
    """
    if type(arg) == list:
        for subarg in arg:
            _check_arg_in_list(subarg, allowed_list, split=split)
        return
    if split is not None:
        for subarg in arg.split(split):
            _check_arg_in_list(subarg, allowed_list)
    if arg not in allowed_list:
        raise ValueError(f"Incorrect argument passed: {arg}")
