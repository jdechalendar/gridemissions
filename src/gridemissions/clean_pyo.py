import pandas as pd
import time
import re
import pyomo.environ as pyo
from pyomo.opt import SolverFactory
from gridemissions import eia_api_v2 as eia
from .clean_v2 import Cleaner, EPSILON
from .load import BaData


class PyoCleaningModel(object):
    """
    Create an AbstractModel() for the cleaning problem.

    No data is passed into this model at this point, it is
    simply written in algebraic form.
    """

    def __init__(self):
        m = pyo.AbstractModel()

        # Sets
        m.regions = pyo.Set()
        m.srcs = pyo.Set()
        m.regions2 = pyo.Set(within=m.regions * m.regions)
        m.regions_srcs = pyo.Set(within=m.regions * m.srcs)

        # Parameters
        m.D = pyo.Param(m.regions, within=pyo.Reals)
        m.NG = pyo.Param(m.regions, within=pyo.Reals)
        m.TI = pyo.Param(m.regions, within=pyo.Reals)
        m.ID = pyo.Param(m.regions2, within=pyo.Reals)
        m.NG_SRC = pyo.Param(m.regions_srcs, within=pyo.Reals)
        m.D_W = pyo.Param(m.regions, default=1.0, within=pyo.Reals)
        m.NG_W = pyo.Param(m.regions, default=1.0, within=pyo.Reals)
        m.TI_W = pyo.Param(m.regions, default=1.0, within=pyo.Reals)
        m.ID_W = pyo.Param(m.regions2, default=1.0, within=pyo.Reals)
        m.NG_SRC_W = pyo.Param(m.regions_srcs, default=1.0, within=pyo.Reals)

        # Variables
        # delta_NG_aux are aux variable for the case where there
        # are no SRC data. In that case, the NG_sum constraint would
        # only have: m.NG + m.delta_NG = 0.
        m.delta_D = pyo.Var(m.regions, within=pyo.Reals)
        m.delta_NG = pyo.Var(m.regions, within=pyo.Reals)
        m.delta_TI = pyo.Var(m.regions, within=pyo.Reals)
        m.delta_ID = pyo.Var(m.regions2, within=pyo.Reals)
        m.delta_NG_SRC = pyo.Var(m.regions_srcs, within=pyo.Reals)
        #         m.delta_NG_aux = pyo.Var(m.regions, within=pyo.Reals)

        # Constraints
        m.D_positive = pyo.Constraint(m.regions, rule=self.D_positive)
        m.NG_positive = pyo.Constraint(m.regions, rule=self.NG_positive)
        m.NG_SRC_positive = pyo.Constraint(m.regions_srcs, rule=self.NG_SRC_positive)
        m.energy_balance = pyo.Constraint(m.regions, rule=self.energy_balance)
        m.antisymmetry = pyo.Constraint(m.regions2, rule=self.antisymmetry)
        m.trade_sum = pyo.Constraint(m.regions, rule=self.trade_sum)
        m.NG_sum = pyo.Constraint(m.regions, rule=self.NG_sum)

        # Objective
        m.total_penalty = pyo.Objective(rule=self.total_penalty, sense=pyo.minimize)

        self.m = m

    def D_positive(self, model, i):
        return (model.D[i] + model.delta_D[i]) >= EPSILON

    def NG_positive(self, model, i):
        return (model.NG[i] + model.delta_NG[i]) >= EPSILON

    def NG_SRC_positive(self, model, k, s):
        return model.NG_SRC[k, s] + model.delta_NG_SRC[k, s] >= EPSILON

    def energy_balance(self, model, i):
        return (
            model.D[i]
            + model.delta_D[i]
            + model.TI[i]
            + model.delta_TI[i]
            - model.NG[i]
            - model.delta_NG[i]
        ) == 0.0

    def antisymmetry(self, model, i, j):
        return (
            model.ID[i, j]
            + model.delta_ID[i, j]
            + model.ID[j, i]
            + model.delta_ID[j, i]
            == 0.0
        )

    def trade_sum(self, model, i):
        return (
            model.TI[i]
            + model.delta_TI[i]
            - sum(
                model.ID[k, l] + model.delta_ID[k, l]
                for (k, l) in model.regions2
                if k == i
            )
        ) == 0.0

    def NG_sum(self, model, i):
        return (
            model.NG[i]
            + model.delta_NG[i]  # + model.delta_NG_aux[i]
            - sum(
                model.NG_SRC[k, s] + model.delta_NG_SRC[k, s]
                for (k, s) in model.regions_srcs
                if k == i
            )
        ) == 0.0

    def total_penalty(self, model):
        return (
            sum(
                (
                    model.D_W[i] * model.delta_D[i] ** 2
                    + model.NG_W[i] * model.delta_NG[i] ** 2
                    #                      + model.delta_NG_aux[i]**2
                    + model.TI_W[i] * model.delta_TI[i] ** 2
                )
                for i in model.regions
            )
            + sum(
                model.ID_W[i, j] * model.delta_ID[i, j] ** 2
                for (i, j) in model.regions2
            )
            + sum(
                model.NG_SRC_W[i, s] * model.delta_NG_SRC[i, s] ** 2
                for (i, s) in model.regions_srcs
            )
        )


class PyoCleaner(Cleaner):
    """
    Optimization-based cleaning class.

    Uses pyomo to build the model and Gurobi as the default solver.
    """

    def __init__(self, ba_data, weights=None, solver="gurobi"):
        super().__init__(ba_data)
        self.m = PyoCleaningModel().m
        self.opt = SolverFactory(solver)
        self.weights = weights
        if weights is not None:
            self.d.df = pd.concat(
                [self.d.df, weights.rename(lambda x: x + "_W", axis=1)], axis=1
            )

    def process(self, debug=False):
        start = time.time()
        self.logger.info("Running PyoCleaner for %d rows" % len(self.d.df))
        self.d.df = self.d.df.fillna(0)
        if not debug:
            self.r = self.d.df.apply(self._process, axis=1)
        else:
            r_list = []
            delta_list = []
            for idx, row in self.d.df.iterrows():
                _, r, deltas = self._process(row, debug=True)
                r_list.append(r)
                delta_list.append(deltas)
            self.r = pd.concat(r_list, axis=1).transpose()
            self.deltas = pd.concat(delta_list, axis=1).transpose()
            self.deltas.index = self.d.df.index

        self.r.index = self.d.df.index

        # Make sure the cleaning step performed as expected
        self.r = BaData(df=self.r)
        self.logger.info("Checking BAs...")
        for ba in self.r.regions:
            self.r.checkBA(ba)
        self.logger.info("Execution took %.2f seconds" % (time.time() - start))

    def _process(self, row, debug=False):
        if row.isna().sum() > 0:
            raise ValueError("Cannot call this method on data with NaNs")
        i = self._create_instance(row)
        self.opt.solve(i)

        r = pd.concat(
            [
                pd.Series(
                    {
                        self.d.KEY["NG"] % k: (i.NG[k] + pyo.value(i.delta_NG[k]))
                        for k in i.regions
                    }
                ),
                pd.Series(
                    {
                        self.d.KEY["D"] % k: (i.D[k] + pyo.value(i.delta_D[k]))
                        for k in i.regions
                    }
                ),
                pd.Series(
                    {
                        self.d.KEY["TI"] % k: (i.TI[k] + pyo.value(i.delta_TI[k]))
                        for k in i.regions
                    }
                ),
                pd.Series(
                    {
                        self.d.KEY["ID"]
                        % (k1, k2): (i.ID[k1, k2] + pyo.value(i.delta_ID[k1, k2]))
                        for (k1, k2) in i.regions2
                    }
                ),
                pd.Series(
                    {
                        self.d.KEY[s]
                        % k: (i.NG_SRC[k, s] + pyo.value(i.delta_NG_SRC[k, s]))
                        for (k, s) in i.regions_srcs
                    }
                ),
            ]
        )

        deltas = pd.concat(
            [
                pd.Series(
                    {
                        self.d.KEY["NG"] % k: (pyo.value(i.delta_NG[k]))
                        for k in i.regions
                    }
                ),
                pd.Series(
                    {self.d.KEY["D"] % k: (pyo.value(i.delta_D[k])) for k in i.regions}
                ),
                pd.Series(
                    {
                        self.d.KEY["TI"] % k: (pyo.value(i.delta_TI[k]))
                        for k in i.regions
                    }
                ),
                pd.Series(
                    {
                        self.d.KEY["ID"] % (k1, k2): (pyo.value(i.delta_ID[k1, k2]))
                        for (k1, k2) in i.regions2
                    }
                ),
                pd.Series(
                    {
                        self.d.KEY[s] % k: (pyo.value(i.delta_NG_SRC[k, s]))
                        for (k, s) in i.regions_srcs
                    }
                ),
            ]
        )

        if not debug:
            return r

        return i, r, deltas

    def _create_instance(self, row):
        def append_W(x):
            return [c + "_W" for c in x]

        NG_SRC_data = self._get_ng_src(row)
        NG_SRC_data_W = self._get_ng_src(row, weights=True)
        opt_data = {
            None: {
                "regions": {None: self.d.regions},
                "srcs": {None: eia.FUELS},
                "regions2": {
                    None: list(
                        set(
                            [
                                (re.split(r"\.|-|_", el)[1], re.split(r"\.|-|_", el)[2])
                                for el in self.d.df.columns
                                if "ID" in re.split(r"\.|-|_", el)
                            ]
                        )
                    )
                },
                "regions_srcs": {None: list(NG_SRC_data.keys())},
                "D": self._reduce_cols(row.loc[self.d.get_cols(field="D")].to_dict()),
                "NG": self._reduce_cols(row.loc[self.d.get_cols(field="NG")].to_dict()),
                "TI": self._reduce_cols(row.loc[self.d.get_cols(field="TI")].to_dict()),
                "ID": self._reduce_cols(
                    row.loc[self.d.get_cols(field="ID")].to_dict(), nfields=2
                ),
                "NG_SRC": NG_SRC_data,
            }
        }

        if self.weights is not None:
            opt_data[None]["D_W"] = self._reduce_cols(
                row.loc[append_W(self.d.get_cols(field="D"))].to_dict()
            )
            opt_data[None]["NG_W"] = self._reduce_cols(
                row.loc[append_W(self.d.get_cols(field="NG"))].to_dict()
            )
            opt_data[None]["TI_W"] = self._reduce_cols(
                row.loc[append_W(self.d.get_cols(field="TI"))].to_dict()
            )
            opt_data[None]["ID_W"] = self._reduce_cols(
                row.loc[append_W(self.d.get_cols(field="ID"))].to_dict(), nfields=2
            )
            opt_data[None]["NG_SRC_W"] = NG_SRC_data_W

        instance = self.m.create_instance(opt_data)
        return instance

    def _reduce_cols(self, mydict, nfields=1):
        """
        Helper function to simplify the names in a dictionary
        """
        newdict = {}
        for k in mydict:
            if nfields == 1:
                newk = re.split(r"\.|-|_", k)[1]
            elif nfields == 2:
                newk = (re.split(r"\.|-|_", k)[1], re.split(r"\.|-|_", k)[2])
            else:
                raise ValueError("Unexpected argument")
            newdict[newk] = mydict[k]
        return newdict

    def _get_ng_src(self, r, weights=False):
        """
        Helper function to get the NG_SRC data.
        """
        mydict = {}
        for ba in self.d.regions:
            for fuel in eia.FUELS:
                col = self.d.KEY[fuel] % ba
                if weights:
                    col += "_W"
                if col in self.d.df.columns:
                    mydict[(ba, fuel)] = r[col]
        return mydict
