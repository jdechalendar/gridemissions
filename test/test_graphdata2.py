import gridemissions as ge
from .eia_samples import get_path

ge.configure_logging("DEBUG")


def test_read_csv(snapshot):
    # This test also serves to check metadata for the dataset at different dates
    # For example, two balancing areas were retired between the first set of dates
    # and the second one

    # The file below was downloaded with:
    # start = "2021-07-10"
    # end = "2021-07-20"
    # df = scrape(start, end)
    # df.to_csv(f"E_raw_{start}_{end}.csv")

    for i, filename in enumerate(
        [
            "E_raw_2021-07-10_2021-07-20.csv",
            "E_raw_2023-06-10_2023-06-20.csv",
        ]
    ):
        data = ge.read_csv(get_path(filename))
        assert len(data.df) == 241
        assert len(data.regions) == snapshot(name=f"Number of regions [{i}]")
        assert data.regions == snapshot(name=f"Regions [{i}]")
        assert data.region_fields == snapshot(name=f"Region fields [{i}]")
        assert data.link_fields == snapshot(name=f"Link fields [{i}]")
        assert data.fields == snapshot(name=f"Fields [{i}]")

        data = ge.read_csv(get_path("E_raw_2023-06-10_2023-06-20.csv"))
