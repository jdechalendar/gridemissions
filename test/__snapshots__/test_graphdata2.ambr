# serializer version: 1
# name: test_read_csv[Fields [0]]
  list([
    'COL',
    'D',
    'GAS',
    'NG',
    'TI',
    'WND',
    'OTH',
    'WAT',
    'SUN',
    'NUC',
    'OIL',
    'ID',
  ])
# ---
# name: test_read_csv[Fields [1]]
  list([
    'COL',
    'D',
    'GAS',
    'NG',
    'TI',
    'WND',
    'OTH',
    'SUN',
    'WAT',
    'NUC',
    'OIL',
    'ID',
  ])
# ---
# name: test_read_csv[Link fields [0]]
  list([
    'ID',
  ])
# ---
# name: test_read_csv[Link fields [1]]
  list([
    'ID',
  ])
# ---
# name: test_read_csv[Number of regions [0]]
  72
# ---
# name: test_read_csv[Number of regions [1]]
  70
# ---
# name: test_read_csv[Region fields [0]]
  list([
    'COL',
    'D',
    'GAS',
    'NG',
    'TI',
    'WND',
    'OTH',
    'WAT',
    'SUN',
    'NUC',
    'OIL',
  ])
# ---
# name: test_read_csv[Region fields [1]]
  list([
    'COL',
    'D',
    'GAS',
    'NG',
    'TI',
    'WND',
    'OTH',
    'SUN',
    'WAT',
    'NUC',
    'OIL',
  ])
# ---
# name: test_read_csv[Regions [0]]
  list([
    'AEC',
    'MISO',
    'SOCO',
    'AECI',
    'SPA',
    'SWPP',
    'TVA',
    'AVA',
    'BPAT',
    'CHPD',
    'GCPD',
    'IPCO',
    'NWMT',
    'PACW',
    'AVRN',
    'AZPS',
    'CISO',
    'IID',
    'LDWP',
    'PACE',
    'PNM',
    'SRP',
    'TEPC',
    'WACM',
    'WALC',
    'BANC',
    'TIDC',
    'BCHA',
    'DOPD',
    'GRID',
    'NEVP',
    'PGE',
    'PSEI',
    'SCL',
    'TPWR',
    'CEN',
    'CPLE',
    'DUK',
    'PJM',
    'SCEG',
    'SC',
    'YAD',
    'CPLW',
    'DEAA',
    'SEPA',
    'EPE',
    'ERCO',
    'FMPP',
    'FPC',
    'FPL',
    'JEA',
    'TEC',
    'GVL',
    'SEC',
    'TAL',
    'HST',
    'GRIF',
    'GWA',
    'HGMA',
    'ISNE',
    'HQT',
    'NBSO',
    'NYIS',
    'LGEE',
    'GLHB',
    'IESO',
    'MHEB',
    'AESO',
    'WAUW',
    'WWA',
    'PSCO',
    'SPC',
  ])
# ---
# name: test_read_csv[Regions [1]]
  list([
    'AECI',
    'MISO',
    'SPA',
    'SWPP',
    'TVA',
    'AVA',
    'BPAT',
    'CHPD',
    'GCPD',
    'IPCO',
    'NWMT',
    'PACW',
    'AVRN',
    'AZPS',
    'CISO',
    'IID',
    'LDWP',
    'PACE',
    'PNM',
    'SRP',
    'TEPC',
    'WACM',
    'WALC',
    'BANC',
    'TIDC',
    'BCHA',
    'DOPD',
    'GRID',
    'NEVP',
    'PGE',
    'PSEI',
    'SCL',
    'TPWR',
    'CEN',
    'CPLE',
    'DUK',
    'PJM',
    'SCEG',
    'SC',
    'YAD',
    'CPLW',
    'DEAA',
    'SEPA',
    'SOCO',
    'EPE',
    'ERCO',
    'FMPP',
    'FPC',
    'FPL',
    'JEA',
    'TEC',
    'GVL',
    'SEC',
    'TAL',
    'HST',
    'GRIF',
    'GWA',
    'HGMA',
    'ISNE',
    'HQT',
    'NBSO',
    'NYIS',
    'LGEE',
    'IESO',
    'MHEB',
    'AESO',
    'WAUW',
    'WWA',
    'PSCO',
    'SPC',
  ])
# ---
